libsolidity/analysis/ControlFlowRevertPruner.cpp

```diff
@@ -0,0 +1,159 @@
+		collectCalls(*pair.function, pair.contract);
+
+	// In the second iteration we process anything left pending
+	removeRevertingNodes();
+}
+
+FunctionDefinition const* ControlFlowRevertPruner::resolveCall(FunctionCall const& _functionCall, ContractDefinition const* _contract)
+{
+	auto const& functionType = dynamic_cast<FunctionType const&>(
+		*_functionCall.expression().annotation().type
+	);
+
+	if (!functionType.hasDeclaration())
+		return nullptr;
+
+	auto const& unresolvedFunctionDefinition =
+		dynamic_cast<FunctionDefinition const&>(functionType.declaration());
+
+	if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+	{
+		if (*memberAccess->annotation().requiredLookup == VirtualLookup::Super)
+		{
+			if (auto const typeType = dynamic_cast<TypeType const*>(memberAccess->expression().annotation().type))
+				if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+				{
+					solAssert(contractType->isSuper(), "");
+					ContractDefinition const* superContract = contractType->contractDefinition().superContract(*_contract);
+
+					return &unresolvedFunctionDefinition.resolveVirtual(
+						*_contract,
+						superContract
+					);
+				}
+		}
+		else
+		{
+			solAssert(*memberAccess->annotation().requiredLookup == VirtualLookup::Static, "");
+			return &unresolvedFunctionDefinition;
+		}
+	}
+	else if (auto const* identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+	{
+		solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
+		return &unresolvedFunctionDefinition.resolveVirtual(*_contract);
+	}
+
+	return &unresolvedFunctionDefinition;
+}
+
+void ControlFlowRevertPruner::removeRevertingNodes()
+{
+	std::set<CFG::FunctionContractTuple> pending_functions = keys(m_functions);
+
+	while (!pending_functions.empty())
+	{
+		CFG::FunctionContractTuple item = *pending_functions.begin();
+		pending_functions.erase(pending_functions.begin());
+
+		auto& revertState = m_functions[item];
+
+		auto previousState = revertState;
+		revertState = RevertState::AllPathsRevert;
+
+		FunctionFlow const& functionFlow = m_cfg.functionFlow(*item.function, item.contract);
+
+		solidity::util::BreadthFirstSearch<CFGNode*>{{functionFlow.entry}}.run(
+			[&](CFGNode* _node, auto&& _addChild) {
+				if (_node == functionFlow.exit)
+					revertState = RevertState::HasNonRevertingPath;
+
+				for (auto const* functionCall: _node->functionCalls)
+				{
+					auto const* resolvedFunction = resolveCall(*functionCall, item.contract);
+
+					if (
+						resolvedFunction == nullptr ||
+						resolvedFunction == item.function ||
+						!resolvedFunction->isImplemented()
+					)
+						continue;
+
+					ContractDefinition const* mostDerivedContract = nullptr;
+					if (auto const* functionContract = resolvedFunction->annotation().contract)
+					{
+						if (item.contract && item.contract->derivesFrom(*functionContract))
```

Comment:
Please add a comment: When calling a base function, we keep the most derived, but we use the called contract in case it is a library function or a free function.

Reply:
updated. Please add the comment above the declaration of `mostDerivedContract`.